### DISCUSSION TITLE
Good idea to use LTSP to set up network booted worker nodes in a small research cluster?

### DISCUSSION THREAD

**Discussion by ertzuiop** [ltsp/ltsp#160] [Q&A]
Hello everyone,

I am new to LTSP and I have a more general question, since I am not sure if LTSP is the right tool for my purpose. Here is what I am planning to do:

- set up a small cluster for computations in my research group (~6 machines, all have the same hardware except for the GPUs)
- Operating system: Ubuntu 20.04 LTS
- the machines have two ethernet ports and will communicate with each other over their own "cluster network" while users login over the other network
- I want to use SLURM to efficiently distribute calculation jobs submitted by users to the worker nodes
- no "HPC" capabilities are planned (meaning: no shared memory between nodes, MPI / parallel software running on multiple nodes. All calculation jobs will be single node processes) 
- users, groups and available software need to be synchronized across the cluster, which means that my worker nodes will be essentially clones of my head node with some small changes to their configuration and running services (in particular: the head node should run the SLURM control, database and job daemons, while the worker nodes should only run the job daemon. I think that the LTSP functionality to mask system services should be suited to achieve this).

I put some thought on how to minimize maintenance effort and how to reduce the risk for the machines to get out of sync (e.g. new software or software updates are installed, but not on all machines, users ids are not synchronized across the cluster, software configuration is not synchronized...).  After doing a lot of googling I came to the conclusion that LTSP fat clients may be what I want. I hope that with these:

- I can setup and configure my head node like I would if it was only a single server
- I can quickly update the worker node image and reboot the workers if updates or additional software is installed on the head node that requires a reboot 
- I can put non-system / application software on a network share mounted by all workers. If application software is installed or updated, no reboot is required, since it is immediately available on all nodes via the network share (e.g. /usr/local is on a local drive of the head node but exported via NFS over the cluster network and mounted by all workers to their /usr/local)
- If can add additional nodes easily by configuring them to boot the image

Unfortunately, I did not find any description from someone using LTSP fat clients to setup cluster worker nodes. So I am wondering if I am missing something and it doesn't really fit my purpose.

So my questions are:

- Is this a good idea at all? Would you recommend using LTSP for this? Are there any pitfalls or limitations that come to your mind with regard to my intended use of LTSP?
- Has someone here ever done something similar?
- Would you rather recommend an alternative method? My worker nodes do have hard drives, so it is not mandatory that they boot from the network, as long as there is a painless way to keep them synchronized. I am currently considering Fully Automed Installation (FAI) + a configuration management tool like Ansible or Puppet, but these tools appear pretty complicated to me and unfortunately I have no experience with these, either.

Any opinion or comment on this is greatly appreciated. Please ask if you need additional information on my use case and requirements.

Thanks a lot in advance!

**Comment by alkisg** [comment]
Hello ertzuiop,

people have used LTSP as cluster nodes in the past, but personally I haven't.
I think it's a perfect match. Indeed, the reason we're using LTSP in Greek schools instead of ansible or puppet, is its incredible ease of maintenance.

The only part that needs to worry you is this:
LTSP automates the centralized management of user accounts and /home.
It also automates things that can be configured from [ltsp.conf](https://ltsp.org/man/ltsp.conf), like HOSTNAME, FSTAB, Xorg...

Now suppose that when a node joins a cluster, something in /etc/some/file gets modified, and it is supposed to be different for each node.
In that case, you'll need to discover which file needs to be different per node, what should be its contents, and then you should make a tiny ltsp script in /usr/share/ltsp/client/init that puts the appropriate contents in the appropriate file when the client boots.

**Comment by ertzuiop** [comment]
This sounds like LTSP is exactly what I was looking for. Thank you very much!

**Comment by helder-daniel** [comment]
Good afternoon, have you made any progress in building your cluster using ltsp? I'm also interested in building one! Any help will be welcome! Thanks!

**Reply by abdeoliveira** [reply]
I managed to setup a similar cluster as described in this post with small variations. So far so good during the first month of experiment. Please contact me if you need assistance on building yours.
